src/client/ds3/object.c

```diff
@@ -199,21 +204,51 @@ ds3_obj_set_info(struct ds3_object_info *info, ds3_bucket_t *ds3b, ds3_obj_t *ds
 			     info->encoded_length, 0);
 }
 
+static int
+ds3_obj_read_int_cb(void *args, daos_event_t *ev, int ret)
+{
+	D_FREE(args);
+	return 0;
+}
+
+static int
+ds3_obj_read_int(void *buf, daos_off_t off, daos_size_t *size, ds3_bucket_t *ds3b, ds3_obj_t *ds3o,
+	     daos_event_t *ev)
+{
+	ds3_obj_args_t	*args;
+
+	D_ALLOC_PTR(args);
+	if (args == NULL)
+		return -DER_NOMEM;
+
+	d_iov_set(&args->iov, buf, *size);
+	args->sg.sg_nr     = 1;
+	args->sg.sg_iovs   = &args->iov;
+	args->sg.sg_nr_out = 1;
+
+	daos_event_register_comp_cb(ev, ds3_obj_read_int_cb, args);
+	return -dfs_read(ds3b->dfs, ds3o->dfs_obj, &args->sg, off, size, ev);
+}
+
 int
 ds3_obj_read(void *buf, daos_off_t off, daos_size_t *size, ds3_bucket_t *ds3b, ds3_obj_t *ds3o,
 	     daos_event_t *ev)
 {
-	d_iov_t     iov;
-	d_sg_list_t rsgl;
-
 	if (ds3b == NULL || buf == NULL || ds3o == NULL)
 		return -EINVAL;
 
-	d_iov_set(&iov, buf, *size);
-	rsgl.sg_nr     = 1;
-	rsgl.sg_iovs   = &iov;
-	rsgl.sg_nr_out = 1;
-	return -dfs_read(ds3b->dfs, ds3o->dfs_obj, &rsgl, off, size, ev);
+	if (ev == NULL) {
```

Comment:
Alternatively, `ev` NULL check can be moved to `ds3_obj_read_int()` before invoking `daos_event_register_comp_cb()` or merge `ds3_obj_read_int()` and `ds3_obj_read()`?

Reply:
I implemented it with a separate function for the async case as that is what `dfs_read` does. But I'm fine combining the functions if that is preferred.

Reply:
Since we are trying to avoid stack allocated variables here, may be we should do that unconditionally with respect to `ev` 's presence.
Presently we are allocating this structure only in the async case and rgw is not using async path as of now.
We can allocate `ds3_obj_args_t` for both sync and async by invoking `ds3_obj_read_int()` and configuring the callback if `ev`'s not NULL. That way we can avoid code duplication and have the same code path.

It would have been great to have altogether separate functions for sync and async from readability perspective,
e.g. `ds3_obj_read(void *buf, daos_off_t off, daos_size_t *size, ds3_bucket_t *ds3b, ds3_obj_t *ds3o)` and `ds3_obj_read_async(void *buf, daos_off_t off, daos_size_t *size, ds3_bucket_t *ds3b, ds3_obj_t *ds3o,
	     daos_event_t *ev)`
Check `Java_io_daos_dfs_DaosFsClient_dfsRead()` and `Java_io_daos_dfs_DaosFsClient_dfsReadAsync()`. 
But let's not do that here.
I see different conventions at client and dfs levels though, may be this is fine, but then we must not use stack in either case for consistency, I think. I will leave it up to you.

Reply:
no we should not do allocation in case of the blocking case, just to avoid the unecessary overhead IMO.

Reply:
I am okay, since we are explicitly just handling the async case here. @mchaarawi, another question I had is, I believe `dfs_read()` would return immediately being async, the actual operation result would be received in the call back's `ret` param, right? we are not really using that or propagating that in this case, what do you think? or is it that we need to handle it separately?

Reply:
> no we should not do mem allocation in case of the blocking case, just to avoid the unnecessary overhead IMO.

Would it still make sense to check for a NULL `ev` in `ds3_obj_read_int()`? If so, would the following be correct?
```diff
diff --git a/src/client/ds3/object.c b/src/client/ds3/object.c
index fd9e7c838..f80fd5dfb 100644
--- a/src/client/ds3/object.c
+++ b/src/client/ds3/object.c
@@ -217,6 +217,9 @@ ds3_obj_read_int(void *buf, daos_off_t off, daos_size_t *size, ds3_bucket_t *ds3
 {
        ds3_obj_args_t  *args;

+       D_ASSERT(ev);
+       daos_event_errno_rc(ev);
+
        D_ALLOC_PTR(args);
        if (args == NULL)
                return -DER_NOMEM;
```